### PR TITLE
dev/ci: remove stateless runtype override, enable caching

### PR DESCRIPTION
**Pull request by bobheadxi**
Before we merge this, we must disable the stateless agent testing pipeline to avoid duplicated main runs. From what I understand this will allow us to properly use the caching features

## Test plan

https://github.com/sourcegraph/sourcegraph/pull/32751#issuecomment-1072521750

### CODE REVIEW

**Comment by jhchabran** [comment]
We need to change how the stateless agent flag is computed, then because it's not based on which queue (though that should have been how it should have been) but instead rely on an env var. 

When we're retrying builds on that queue, if we don't have this flag, we won't rely on the cache. I'll add some commits to reflect that.

**Comment by bobheadxi** [comment]
I've removed the schedule from https://buildkite.com/sourcegraph/sourcegraph-stateless/settings/schedules as well

**Comment by bobheadxi** [comment]
https://buildkite.com/sourcegraph/sourcegraph/builds/137832 is a test re-build that will target stateless agents

File: enterprise/dev/ci/internal/buildkite/buildkite.go [hunk 1]
```diff
@@ -30,7 +30,10 @@ type featureFlags struct {
 
 // FeatureFlags are for experimenting with CI pipeline features. Use sparingly!
 var FeatureFlags = featureFlags{
-	StatelessBuild: os.Getenv("CI_FEATURE_FLAG_STATELESS") == "true",
+	StatelessBuild: os.Getenv("CI_FEATURE_FLAG_STATELESS") == "true" ||
+		// Always process retries on stateless agents.
+		// TODO: remove when we switch over entirely to stateless agents
+		os.Getenv("BUILDKITE_REBUILT_FROM_BUILD_NUMBER") != "",
```
Comment (bobheadxi) [inline enterprise/dev/ci/internal/buildkite/buildkite.go:36]: @jhchabran I had to revert this, this isn't computed per-job but per-build - the "stateless" state is evaluated pre-pipeline-generation. So either the entire build targets stateless, or we get assigned onto a machine at random

File: enterprise/dev/ci/internal/buildkite/buildkite.go [hunk 2]
```diff
@@ -30,7 +31,12 @@ type featureFlags struct {
 
 // FeatureFlags are for experimenting with CI pipeline features. Use sparingly!
 var FeatureFlags = featureFlags{
-	StatelessBuild: os.Getenv("CI_FEATURE_FLAG_STATELESS") == "true",
+	StatelessBuild: os.Getenv("CI_FEATURE_FLAG_STATELESS") == "true" ||
```
Comment (bobheadxi) [inline enterprise/dev/ci/internal/buildkite/buildkite.go:34]: @jhchabran I had to revert this, this isn't computed per-job but per-build - the "stateless" state is evaluated pre-pipeline-generation. So either the entire build targets stateless, or we get assigned onto a machine at random

**Comment by bobheadxi** [comment]
https://buildkite.com/sourcegraph/sourcegraph/builds/137837#897ba3ef-78f9-47f8-9859-43471d760848 is a build targetting `queue=job`

File: enterprise/dev/ci/internal/buildkite/buildkite.go (hunk 2 above)
Comment (bobheadxi) [reply enterprise/dev/ci/internal/buildkite/buildkite.go:34]: e.g. https://buildkite.com/sourcegraph/sourcegraph/builds/137834#14ec74bc-3bb3-4ffe-9b6b-3cd16e392eab indicates "skipping yarn cache", even though it was `queue=job`

Comment (jhchabran) [reply enterprise/dev/ci/internal/buildkite/buildkite.go:34]: Oh right, I remember now this is how it works 🙏

Comment (bobheadxi) [reply enterprise/dev/ci/internal/buildkite/buildkite.go:34]: Yeah this got me yesterday as well 😆 Everything is evaluated based on the lucky agent that gets the pipeline generation job

File: enterprise/dev/ci/internal/buildkite/agents.go [hunk 3]
```diff
@@ -3,6 +3,6 @@ package buildkite
 const (
 	AgentQueueStandard  = "standard"
 	AgentQueueBaremetal = "baremetal"
-	AgentQueueJob       = "job"
+	AgentQueueStateless = "stateless"
```
Comment (bobheadxi) [inline enterprise/dev/ci/internal/buildkite/agents.go:6]: @davejrt are you updating this on the agents as well?

File: enterprise/dev/ci/internal/buildkite/buildkite.go [hunk 4]
```diff
@@ -30,7 +31,12 @@ type featureFlags struct {
...
+		// Always process retries on stateless agents.
+		// TODO: remove when we switch over entirely to stateless agents
+		os.Getenv("BUILDKITE_REBUILT_FROM_BUILD_NUMBER") != "" ||
+		// Roll out to 75% of builds
+		rand.Uint32()%100 < uint32(75),
```
[the remark this answers is not in the document]
Comment (jhchabran) [reply enterprise/dev/ci/internal/buildkite/buildkite.go:39]: I think we need to use cryptorand there

**Comment by bobheadxi** [comment]
Stateless rebuild: https://buildkite.com/sourcegraph/sourcegraph/builds/137854

**Review by davejrt — APPROVED** [review]

**Review by jhchabran — APPROVED** [review]

**Pull request merged**